cpp/include/cudf/utilities/traits.hpp

```diff
@@ -486,3 +486,18 @@ constexpr inline bool is_nested(data_type type)
 
 /** @} */
 }  // namespace cudf
+
+namespace std {
+// These specializations are needed due to nvcc bugs. The first one can work in 10.2
```

Comment:
This is undefined behavior:
```
The behavior of a program that adds specializations for is_trivially_copyable or is undefined.
```
https://en.cppreference.com/w/cpp/types/is_trivially_copyable

In general it's UB to specialized `std::` traits.

Reply:
the object layout and the result of using/inspecting the trait isn't changing. in fact, this is making it return the correct values (as though the specialization had not been there), as the existing version is saying some conversions aren't possible, when they in fact are. 

i understand what the standard is saying, and hopefully these may no longer be required with a higher cuda version/compiler. what specifically do you think will make this undefined behavior in this context? is it just the fact the standards says no, and we don't.

Reply:
> what specifically do you think will make this undefined behavior in this context? 

Yes, because the standard literally says it is undefined behavior.

Reply:
What led you to wanting to do this in the first place?

Reply:
> allows widening of timestamp type - i.e. to say, allows timestamp with a lower precision to be convertible to timestamp with a higher precision and not vice-versa.

That sounds like a case where `time_point_cast` should be used, not implicit conversion.

Reply:
- there exists a constructor in time_point today that does allow this implicit construction without a cast
- the is_convertible type trait is used today and where used, such conversions can be supported naturally

Reply:
> there exists a constructor in time_point today that does allow this implicit construction without a cast

My suspicion is that this is a bug. 

> the is_convertible type trait is used today and where used, such conversions can be supported naturally

But they can't be supported naturally:
1. It overrides the type rules. Those rules exist for good reason and we should follow them.
2. Specializing the `std::is_convertible` trait results in undefined behavior.

Reply:
>> My suspicion is that this is a bug.

do you mean to say that `std::chrono` should not have allowed this? neither the documentation for time_point nor duration have any caveats sorrounding this. fwiw, i tried to compile this with gcc-11.x (presumably latest), and it works

```
    timestamp_s secs{std::chrono::duration<int64_t, ::std::chrono::seconds::period>{10}};
    timestamp_ms msecs{std::chrono::duration<int64_t, ::std::chrono::milliseconds::period>{10000}};
    timestamp_s msecs2 = secs;
    //timestamp_s secs2 = msecs; // error
```

i get that specializing traits isn't recommended and that there should be no reason to override these type rules. the traits specialization are done here only to support/include conversions that are allowed, yet aren't due to compiler issues.

if you vehemently oppose this and do not want to consider this as a stop gap, i'll revert it.

Reply:
I'm having difficult understanding what exactly you are trying to do.

So to be abundantly clear:

https://godbolt.org/z/dTafZ4

`std::chrono` allows converting `time_point`s from low to high, but not high to low. 

So what is the problem in libcudf that you are trying to solve?

Reply:
i'm trying to do [this](https://github.com/rapidsai/cudf/pull/5394#discussion_r438808700). if i do the first part of it, then the `is_convertible` type_trait is saying all conversions are allowed (with nvcc), and thus the specialization. this trait is used [here](https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/src/reductions/simple.cuh#L80) with timestamps.

Reply:
Discussed with @sriramch offline and the issue is summarized here: https://godbolt.org/z/bK-eJp

nvcc is mistakenly allowing conversion from a higher precision time_point to a lower precision time_point. nvbug is filed.

Reply:
Can we add a chrono specialization overload for [`result_type_dispatcher::is_supported_v`](https://github.com/rapidsai/cudf/blob/9850c551a89c34d10e30f33965cc987b4e1e9f8c/cpp/src/reductions/simple.cuh#L73) instead of customizing the std traits?

Reply:
that's a great idea. however, if `std::is_convertible` doesn't work in the future for another pair of types, that place may need a customization as well. by keeping this in a central place and specializing it on a case by case basis, we can quickly revert the cudf implementation to the `std` type_traits when a later version of the compiler is supported.

Reply:
> we can quickly revert the cudf implementation to the std type_traits when a later version of the compiler is supported

But not until we entirely stop building for CUDA 10.0-10.2, which won't be for a while.

I agree with @jrhemstad here:
> Solution is to create and use our on is_convertible trait that forwards to std::is_convertible except for chrono types where we can inject our own logic until the nvcc issue is resolved.

Reply:
there is an issue with `is_trivially_copyable` as well with nvcc 10.0. it doesn't look like cudf wrappers could be created for this trait, as this is enforced within rmm. will the cudf 10.0 builds be removed from build pipeline for 0.15? i may have to keep the stopgap specialization of this trait in the std namespace to get the builds going for 10.0 in the interim.

Reply:
Is it only in 10.0? Not 10.1?

Reply:
yes!